File: Documentation/5.4/Samples/python/ClientApi/Session/Querying/MoreLikeThis.py
from typing import List, Union, Callable, TypeVar

from ravendb import (
    DocumentQuery,
    MoreLikeThisBase,
    MoreLikeThisBuilder,
    MoreLikeThisOperations,
    MoreLikeThisOptions,
    AbstractIndexCreationTask,
)

from examples_base import ExampleBase

_T = TypeVar("_T")


class Article:
    def __init__(self, Id: str = None, category: str = None):
        self.Id = Id
        self.category = category


class Articles_MoreLikeThis(AbstractIndexCreationTask):
    def __init__(self):
        super().__init__()
        self.map = 'from o in docs.Articles select new { Id = o.Id, category = o.category, body = "aaaa" }'


class MoreLikeThisExample(ExampleBase):
    def setUp(self):
        super().setUp()

    class Foo:
        # region more_like_this_2
        def __init__(
            self,
            minimum_term_frequency: int = None,
            maximum_query_terms: int = None,
            maximum_number_of_tokens_parsed: int = None,
            minimum_word_length: int = None,
            maximum_word_length: int = None,
            minimum_document_frequency: int = None,
            maximum_document_frequency: int = None,
            maximum_document_frequency_percentage: int = None,
            boost: bool = None,
            boost_factor: float = None,
            stop_words_document_id: str = None,
            fields: List[str] = None,
        ): ...

        # endregion

    class Foo2:
        # region more_like_this_1
        def more_like_this(
            self, more_like_this_or_builder: Union[MoreLikeThisBase, Callable[[MoreLikeThisBuilder[_T]], None]]
        ) -> DocumentQuery[_T]: ...

        # endregion
        # region more_like_this_3
        def using_any_document(self) -> MoreLikeThisOperations[_T]: ...

        def using_document(
            self, document_json_or_builder: Union[str, Callable[[DocumentQuery[_T]], None]]
        ) -> MoreLikeThisOperations[_T]: ...

        def with_options(self, options: MoreLikeThisOptions) -> MoreLikeThisOperations[_T]: ...

        # endregion

    def test_more_like_this(self):
        with self.embedded_server.get_document_store("MoreLikeThisExample") as store:
            Articles_MoreLikeThis().execute(store)
            with store.open_session() as session:
                session.store(Article("articles/1", "IT"))
                session.store(Article("articles/12", "ITl"))
                session.store(Article("articles/111", "trITl"))
                session.save_changes()
                # region more_like_this_4
                # Search for similar articles to 'articles/1'
                # using 'Articles/MoreLikeThis' index and search only field 'body'
                articles = list(
                    session.query_index_type(Articles_MoreLikeThis, Article).more_like_this(
                        lambda builder: builder.using_document(
                            lambda x: x.where_equals("Id", "articles/1")
                        ).with_options(MoreLikeThisOptions(fields=["body"]))
                    )
                )
                # endregion

            with store.open_session() as session:
                # region more_like_this_6
                # Search for similar articles to 'articles/1'
                # using 'Articles/MoreLikeThis' index and search only field 'body'
                # where article category is 'IT'
                articles = list(
                    session.query_index_type(Articles_MoreLikeThis, Article)
                    .more_like_this(
                        lambda builder: builder.using_document(
                            lambda x: x.where_equals("Id", "articles/1")
                        ).with_options(MoreLikeThisOptions(fields=["body"]))
                    )
                    .where_equals("category", "IT")
                )
                # endregion

File: Documentation/5.4/Samples/php/ClientApi/Session/Querying/MoreLikeThis.php
<?php

use RavenDB\Constants\PhpClient;
use RavenDB\Documents\Queries\MoreLikeThis\MoreLikeThisBase;
use RavenDB\Documents\Queries\MoreLikeThis\MoreLikeThisOperationsInterface;
use RavenDB\Documents\Queries\MoreLikeThis\MoreLikeThisOptions;
use RavenDB\Documents\Queries\Query;
use RavenDB\Documents\Session\DocumentQueryInterface;

class Foo {
    # region more_like_this_2
    public const DEFAULT_MAXIMUM_NUMBER_OF_TOKENS_PARSED = 5000;
    public const DEFAULT_MINIMUM_TERM_FREQUENCY = 2;
    public const DEFAULT_MINIMUM_DOCUMENT_FREQUENCY = 5;
    public const DEFAULT_MAXIMUM_DOCUMENT_FREQUENCY = PhpClient::INT_MAX_VALUE;
    public const DEFAULT_BOOST = false;
    public const DEFAULT_BOOST_FACTOR = 1;
    public const DEFAULT_MINIMUM_WORD_LENGTH = 0;
    public const DEFAULT_MAXIMUM_WORD_LENGTH = 0;
    public const DEFAULT_MAXIMUM_QUERY_TERMS = 25;

    private ?int $minimumTermFrequency = null;
    private ?int $maximumQueryTerms = null;
    private ?int $maximumNumberOfTokensParsed = null;
    private ?int $minimumWordLength = null;
    private ?int $maximumWordLength = null;
    private ?int $minimumDocumentFrequency = null;
    private ?int $maximumDocumentFrequency = null;
    private ?int $maximumDocumentFrequencyPercentage = null;
    private ?bool $boost = null;
    private ?float $boostFactor = null;
    private ?string $stopWordsDocumentId = null;
    private ?array $fields = null;

    # endregion
}

interface FooInterface2 {
    # region more_like_this_1

    /**
     * Usage:
     *   - moreLikeThis(MoreLikeThisBase $moreLikeThis);
     *   - moreLikeThis(function(MoreLikeThisBuilder($builder) {...});
     *
     * @param MoreLikeThisBase|Closure|null $moreLikeThisOrBuilder
     * @return DocumentQueryInterface
     */
    public function moreLikeThis(null|MoreLikeThisBase|Closure $moreLikeThisOrBuilder): DocumentQueryInterface;
    # endregion

    # region more_like_this_3
    /**
     * Usage:
     *   - usingDocument();
     *   - usingDocument(string $documentJson);
     *   - usingDocument(function(MoreLikeThisBuilder $build) {...});
     *
     * @param string|Closure|null $documentJsonOrBuilder
     * @return MoreLikeThisOperationsInterface
     */
    function usingDocument(null|string|Closure $documentJsonOrBuilder): MoreLikeThisOperationsInterface;

    function usingDocumentWithJson(?string $documentJson): MoreLikeThisOperationsInterface; // same as calling usingDocument(string $documentJson)

    function usingDocumentWithBuilder(?Closure $builder): MoreLikeThisOperationsInterface; // same as calling usingDocument(function($builder) {...});

    function withOptions(MoreLikeThisOptions $options): MoreLikeThisOperationsInterface;
    # endregion
}

class Article {
    private ?string $id = null;
    private ?string $category = null;

    public function getId(): ?string
    {
        return $this->id;
    }

    public function setId(?string $id): void
    {
        $this->id = $id;
    }

    public function getCategory(): ?string
    {
        return $this->category;
    }

    public function setCategory(?string $category): void
    {
        $this->category = $category;
    }
}

class MoreLikeThis
{
    public function samples(): void
    {
        $store = DocumentStoreHolder::getStore();
        try {

            $session = $store->openSession();
            try {
                # region more_like_this_4
                // Search for similar articles to 'articles/1'
                // using 'Articles/MoreLikeThis' index and search only field 'body'
                $options = new MoreLikeThisOptions();
                $options->setFields([ "body" ]);

                /** @var array<Article> $articles */
                $articles = $session
                    ->query(Article::class, Query::index("Articles/MoreLikeThis"))
                    ->moreLikeThis(function($builder) use ($options) {
                        $builder
                            ->usingDocument(function ($x) {
                                $x->whereEquals("id()", "articles/1");
                            })
                            ->withOptions($options);
                    })
                    ->toList();
                # endregion
            } finally {
                $session->close();
            }

            $session = $store->openSession();
            try {
                # region more_like_this_6
                // Search for similar articles to 'articles/1'
                // using 'Articles/MoreLikeThis' index and search only field 'body'
                // where article category is 'IT'
                $options = new MoreLikeThisOptions();
                $options->setFields([ "body" ]);
                /** @var array<Article> $articles */
                $articles = $session
                    ->query(Article::class, Query::index("Articles/MoreLikeThis"))
                    ->moreLikeThis(function($builder) use ($options) {
                        $builder
                            ->usingDocument(function ($x) {
                                $x->whereEquals("id()", "articles/1");
                            })
                            ->withOptions($options);
                    })
                    ->whereEquals("category", "IT")
                    ->toList();
                # endregion
            } finally {
                $session->close();
            }
        } finally {
            $store->close();
        }
    }
}
